File: python/week15/714.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        if len(prices) <= 1:
            return 0
        buy, sell = -prices[0], 0
        for i in prices[1:]:
            buy = max(buy, sell-i)
            sell = max(sell, i-fee+buy)
        return sell

File: go/714.go
func maxProfit(prices []int, fee int) int {
    if len(prices) == 1 {
        return 0
    }
    var buy int = -prices[0]
    var sell int = 0
    for i:= 1; i < len(prices); i++ {
        buy = maxInt(buy, sell-prices[i])
        sell = maxInt(sell, prices[i]-fee+buy)
    }
    return sell
}
func maxInt(a int, b int) int {
    if a < b {
        return b
    }
    return a
}